File: src/hono-server/routes/market/social.ts
import { Hono } from 'hono';

import { MarketService } from '@/server/services/market';

import { jsonError, MARKET_BASE_URL } from './helpers';

const social = new Hono();

social.get('/market/social-profile/claimable-resources', async (c) => {
  const marketService = await MarketService.createFromRequest(c.req.raw);

  try {
    const headers = (marketService.market as any).headers as Record<string, string>;
    const response = await fetch(`${MARKET_BASE_URL}/api/v1/user/claims/scan`, {
      headers,
      method: 'GET',
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      return jsonError(
        c,
        'scan_claimable_resources_failed',
        errorData.error || `Failed to scan claimable resources: ${response.status}`,
        response.status,
      );
    }

    const responseData = await response.json();
    const data = responseData.data || responseData;

    return c.json({
      plugins: data.plugins || [],
      skills: data.skills || [],
    });
  } catch (e: any) {
    console.error('[Market SocialProfile] Scan claimable resources failed:', e);
    return jsonError(c, 'scan_claimable_resources_failed', e?.message || 'Unknown error', 500);
  }
});

social.post('/market/social-profile/claim-resources', async (c) => {
  const marketService = await MarketService.createFromRequest(c.req.raw);

  try {
    const body = (await c.req.json()) as {
      pluginIds?: string[];
      skillIds?: string[];
    };
    const headers = (marketService.market as any).headers as Record<string, string>;

    const claimed: Array<{ assetId: number; assetType: 'plugin' | 'skill' }> = [];
    const errors: string[] = [];

    for (const skillId of body.skillIds || []) {
      try {
        const response = await fetch(`${MARKET_BASE_URL}/api/v1/user/claims`, {
          body: JSON.stringify({ assetId: Number(skillId), assetType: 'skill' }),
          headers: {
            ...headers,
            'Content-Type': 'application/json',
          },
          method: 'POST',
        });

        if (response.ok) {
          claimed.push({ assetId: Number(skillId), assetType: 'skill' });
        } else {
          const error = await response.json().catch(() => ({}));
          errors.push(error.error || `Failed to claim skill ${skillId}`);
        }
      } catch {
        errors.push(`Failed to claim skill ${skillId}`);
      }
    }

    for (const pluginId of body.pluginIds || []) {
      try {
        const response = await fetch(`${MARKET_BASE_URL}/api/v1/user/claims`, {
          body: JSON.stringify({ assetId: Number(pluginId), assetType: 'plugin' }),
          headers: {
            ...headers,
            'Content-Type': 'application/json',
          },
          method: 'POST',
        });

        if (response.ok) {
          claimed.push({ assetId: Number(pluginId), assetType: 'plugin' });
        } else {
          const error = await response.json().catch(() => ({}));
          errors.push(error.error || `Failed to claim plugin ${pluginId}`);
        }
      } catch {
        errors.push(`Failed to claim plugin ${pluginId}`);
      }
    }

    if (claimed.length === 0 && errors.length > 0) {
      return jsonError(c, 'claim_resources_failed', errors[0], 500);
    }

    return c.json({
      claimed,
      errors: errors.length > 0 ? errors : undefined,
      success: claimed.length > 0,
    });
  } catch (e: any) {
    console.error('[Market SocialProfile] Claim resources failed:', e);
    return jsonError(c, 'claim_resources_failed', e?.message || 'Unknown error', 500);
  }
});

social.post('/market/social/*', async (c) => {
  const marketService = await MarketService.createFromRequest(c.req.raw);
  const sdk = marketService.market;
  const action = c.req.path.replace('/market/social/', '').split('/')[0];

  try {
    const body = await c.req.json();
    switch (action) {
      case 'follow': {
        await sdk.follows.follow(body.followingId);
        return c.json({ success: true });
      }
      case 'unfollow': {
        await sdk.follows.unfollow(body.followingId);
        return c.json({ success: true });
      }
      case 'favorite': {
        await sdk.favorites.addFavorite(body.targetType, body.identifier ?? body.targetId);
        return c.json({ success: true });
      }
      case 'unfavorite': {
        await sdk.favorites.removeFavorite(body.targetType, body.identifier ?? body.targetId);
        return c.json({ success: true });
      }
      case 'like': {
        await sdk.likes.like(body.targetType, body.identifier ?? body.targetId);
        return c.json({ success: true });
      }
      case 'unlike': {
        await sdk.likes.unlike(body.targetType, body.identifier ?? body.targetId);
        return c.json({ success: true });
      }
      case 'toggle-like': {
        return c.json(
          await sdk.likes.toggleLike(body.targetType, body.identifier ?? body.targetId),
        );
      }
      default: {
        return jsonError(c, 'not_found', `Unknown action: ${action}`, 404);
      }
    }
  } catch (e: any) {
    console.error('[Market Social] Action failed:', e);
    return jsonError(c, 'action_failed', e?.message || 'Unknown error', 500);
  }
});

social.get('/market/social/*', async (c) => {
  const marketService = await MarketService.createFromRequest(c.req.raw);
  const sdk = marketService.market;
  const fullPath = c.req.path.replace('/market/social/', '');
  const segments = fullPath.split('/');
  const action = segments[0];

  const url = new URL(c.req.url);
  const limit = url.searchParams.get('pageSize') || url.searchParams.get('limit');
  const offset = url.searchParams.get('offset');
  const paginationParams: { limit?: number; offset?: number } = {};
  if (limit) paginationParams.limit = Number(limit);
  if (offset) paginationParams.offset = Number(offset);

  try {
    switch (action) {
      case 'follow-status': {
        return c.json(await sdk.follows.checkFollowStatus(Number(segments[1])));
      }
      case 'following': {
        return c.json(await sdk.follows.getFollowing(Number(segments[1]), paginationParams));
      }
      case 'followers': {
        return c.json(await sdk.follows.getFollowers(Number(segments[1]), paginationParams));
      }
      case 'follow-counts': {
        const userId = Number(segments[1]);
        const [following, followers] = await Promise.all([
          sdk.follows.getFollowing(userId, { limit: 1 }),
          sdk.follows.getFollowers(userId, { limit: 1 }),
        ]);
        return c.json({
          followersCount: (followers as any).totalCount || (followers as any).total || 0,
          followingCount: (following as any).totalCount || (following as any).total || 0,
        });
      }
      case 'favorite-status': {
        const targetType = segments[1] as 'agent' | 'plugin';
        const val = /^\d+$/.test(segments[2]) ? Number(segments[2]) : segments[2];
        return c.json(await sdk.favorites.checkFavorite(targetType, val as number));
      }
      case 'favorites': {
        return c.json(await sdk.favorites.getMyFavorites(paginationParams));
      }
      case 'user-favorites': {
        return c.json(await sdk.favorites.getUserFavorites(Number(segments[1]), paginationParams));
      }
      case 'favorite-agents': {
        return c.json(
          await sdk.favorites.getUserFavoriteAgents(Number(segments[1]), paginationParams),
        );
      }
      case 'favorite-plugins': {
        return c.json(
          await sdk.favorites.getUserFavoritePlugins(Number(segments[1]), paginationParams),
        );
      }
      case 'like-status': {
        const targetType = segments[1] as 'agent' | 'plugin';
        const val = /^\d+$/.test(segments[2]) ? Number(segments[2]) : segments[2];
        return c.json(await sdk.likes.checkLike(targetType, val as number));
      }
      case 'liked-agents': {
        return c.json(await sdk.likes.getUserLikedAgents(Number(segments[1]), paginationParams));
      }
      case 'liked-plugins': {
        return c.json(await sdk.likes.getUserLikedPlugins(Number(segments[1]), paginationParams));
      }
      default: {
        return jsonError(c, 'not_found', `Unknown action: ${action}`, 404);
      }
    }
  } catch (e: any) {
    console.error('[Market Social] Query failed:', e);
    return jsonError(c, 'query_failed', e?.message || 'Unknown error', 500);
  }
});

export default social;

File: python-backend/app/routers/social.py
"""Social router — follow/favorite/like actions via external Market API proxy.

Frontend: src/services/social.rest.ts
Prefix: /api/social

The social features (follow, favorite, like) are managed by the external
LobeHub Market API. This router proxies requests from the frontend REST client
to that external API, forwarding auth headers.
"""

from __future__ import annotations

import logging
import os
from typing import Any, Optional

import httpx
from fastapi import APIRouter, Depends, HTTPException, Request, status
from pydantic import BaseModel

from app.dependencies import get_current_user_id

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/social", tags=["Social"])

MARKET_BASE_URL = os.getenv("MARKET_BASE_URL", "https://market.lobehub.com")
MARKET_TIMEOUT = 30.0


# ── Helpers ──────────────────────────────────────────────────────────

def _auth_headers(request: Request) -> dict[str, str]:
    """Forward authorization headers from the incoming request."""
    headers: dict[str, str] = {}
    auth = request.headers.get("authorization")
    if auth:
        headers["Authorization"] = auth
    cookie = request.headers.get("cookie")
    if cookie:
        headers["Cookie"] = cookie
    return headers


async def _proxy_get(
    path: str,
    params: dict[str, Any] | None = None,
    headers: dict[str, str] | None = None,
) -> Any:
    url = f"{MARKET_BASE_URL}{path}"
    clean_params = {k: v for k, v in (params or {}).items() if v is not None}
    async with httpx.AsyncClient(timeout=MARKET_TIMEOUT) as client:
        resp = await client.get(url, params=clean_params, headers=headers or {})
        if resp.status_code >= 400:
            logger.warning("Market social API error: GET %s -> %d", url, resp.status_code)
            raise HTTPException(resp.status_code, resp.text)
        return resp.json()


async def _proxy_post(
    path: str,
    body: dict[str, Any] | None = None,
    headers: dict[str, str] | None = None,
) -> Any:
    url = f"{MARKET_BASE_URL}{path}"
    async with httpx.AsyncClient(timeout=MARKET_TIMEOUT) as client:
        resp = await client.post(url, json=body or {}, headers=headers or {})
        if resp.status_code >= 400:
            logger.warning("Market social API error: POST %s -> %d", url, resp.status_code)
            raise HTTPException(resp.status_code, resp.text)
        return resp.json()


# ── Schemas ──────────────────────────────────────────────────────────

class FollowBody(BaseModel):
    followingId: int


class TargetBody(BaseModel):
    targetId: Optional[int] = None
    identifier: Optional[str] = None
    targetType: str  # 'agent' | 'plugin' | 'agent-group'


# ── Follow endpoints ────────────────────────────────────────────────

@router.post("/follow")
async def follow(
    body: FollowBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/follow",
        body=body.model_dump(),
        headers=_auth_headers(request),
    )


@router.post("/unfollow")
async def unfollow(
    body: FollowBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/unfollow",
        body=body.model_dump(),
        headers=_auth_headers(request),
    )


@router.get("/follow-status")
async def check_follow_status(
    targetUserId: int,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/follow-status",
        params={"targetUserId": targetUserId},
        headers=_auth_headers(request),
    )


@router.get("/follow-counts")
async def get_follow_counts(
    userId: int,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/follow-counts",
        params={"userId": userId},
        headers=_auth_headers(request),
    )


@router.get("/following")
async def get_following(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/following",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


@router.get("/followers")
async def get_followers(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/followers",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


# ── Favorite endpoints ──────────────────────────────────────────────

@router.post("/favorite")
async def add_favorite(
    body: TargetBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/favorite",
        body=body.model_dump(exclude_none=True),
        headers=_auth_headers(request),
    )


@router.post("/unfavorite")
async def remove_favorite(
    body: TargetBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/unfavorite",
        body=body.model_dump(exclude_none=True),
        headers=_auth_headers(request),
    )


@router.get("/favorite-status")
async def check_favorite_status(
    targetType: str,
    targetIdOrIdentifier: str,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/favorite-status",
        params={"targetType": targetType, "targetIdOrIdentifier": targetIdOrIdentifier},
        headers=_auth_headers(request),
    )


@router.get("/my-favorites")
async def get_my_favorites(
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/my-favorites",
        params={"limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


@router.get("/favorite-agents")
async def get_user_favorite_agents(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/favorite-agents",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


@router.get("/favorite-plugins")
async def get_user_favorite_plugins(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/favorite-plugins",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


# ── Like endpoints ──────────────────────────────────────────────────

@router.post("/like")
async def like(
    body: TargetBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/like",
        body=body.model_dump(exclude_none=True),
        headers=_auth_headers(request),
    )


@router.post("/unlike")
async def unlike(
    body: TargetBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/unlike",
        body=body.model_dump(exclude_none=True),
        headers=_auth_headers(request),
    )


@router.get("/like-status")
async def check_like_status(
    targetType: str,
    targetIdOrIdentifier: str,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/like-status",
        params={"targetType": targetType, "targetIdOrIdentifier": targetIdOrIdentifier},
        headers=_auth_headers(request),
    )


@router.post("/toggle-like")
async def toggle_like(
    body: TargetBody,
    request: Request,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_post(
        "/api/social/toggle-like",
        body=body.model_dump(exclude_none=True),
        headers=_auth_headers(request),
    )


@router.get("/liked-agents")
async def get_user_liked_agents(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/liked-agents",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )


@router.get("/liked-plugins")
async def get_user_liked_plugins(
    userId: int,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    request: Request = None,
    user_id: str = Depends(get_current_user_id),
):
    return await _proxy_get(
        "/api/social/liked-plugins",
        params={"userId": userId, "limit": limit, "offset": offset},
        headers=_auth_headers(request),
    )
